clvm-traits/src/to_clvm.rs

```diff
@@ -1,22 +1,50 @@
 use clvmr::{allocator::NodePtr, Allocator};
+use num_bigint::BigInt;
 
-use crate::{Error, Result};
+use crate::{ClvmEncoder, ToClvmError};
 
-pub trait ToClvm {
-    fn to_clvm(&self, a: &mut Allocator) -> Result<NodePtr>;
+pub trait ToClvm<N> {
+    fn to_clvm(&self, encoder: &mut impl ClvmEncoder<Node = N>) -> Result<N, ToClvmError>;
 }
 
-impl ToClvm for NodePtr {
-    fn to_clvm(&self, _a: &mut Allocator) -> Result<NodePtr> {
+pub trait ToNodePtr {
+    fn to_node_ptr(&self, a: &mut Allocator) -> Result<NodePtr, ToClvmError>;
+}
+
+impl<T> ToNodePtr for T
+where
+    T: ToClvm<NodePtr>,
+{
+    fn to_node_ptr(&self, a: &mut Allocator) -> Result<NodePtr, ToClvmError> {
+        self.to_clvm(a)
+    }
+}
+
+impl ToClvm<NodePtr> for NodePtr {
+    fn to_clvm(
+        &self,
+        _encoder: &mut impl ClvmEncoder<Node = NodePtr>,
+    ) -> Result<NodePtr, ToClvmError> {
         Ok(*self)
     }
 }
 
+pub fn simplify_int_bytes(mut slice: &[u8]) -> &[u8] {
+    while (!slice.is_empty()) && (slice[0] == 0) {
+        if slice.len() > 1 && (slice[1] & 0x80 == 0x80) {
+            break;
+        }
+        slice = &slice[1..];
+    }
+    slice
+}
+
 macro_rules! clvm_primitive {
     ($primitive:ty) => {
-        impl ToClvm for $primitive {
-            fn to_clvm(&self, a: &mut Allocator) -> Result<NodePtr> {
-                a.new_number((*self).into()).map_err(Error::Allocator)
+        impl<N> ToClvm<N> for $primitive {
+            fn to_clvm(&self, encoder: &mut impl ClvmEncoder<Node = N>) -> Result<N, ToClvmError> {
+                let number = BigInt::from(*self);
```

Comment:
since all primitives have fixed size, this could be made a lot more efficient by not involving `BigInt`

Reply:
The issue is that unsigned integers may need an additional byte to represent that they're positive in two's complement

Reply:
the max size is still known at compile time though